File: C++/binary_search.cpp
#include <iostream>
using namespace std;

int binarySearch(int arr[], int n, int key) {
    int low = 0, high = n - 1;

    while (low <= high) {
        int mid = (low + high) / 2;  // middle index

        if (arr[mid] == key) {
            return mid; // key found
        }
        else if (arr[mid] < key) {
            low = mid + 1; // search right half
        }
        else {
            high = mid - 1; // search left half
        }
    }

    return -1; // key not found
}

int main() {
    int n, key;

    cout << "Enter number of elements: ";
    cin >> n;

    int arr[n];
    cout << "Enter " << n << " sorted elements: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    cout << "Enter element to search: ";
    cin >> key;

    int result = binarySearch(arr, n, key);

    if (result != -1)
        cout << "Element found at index " << result << endl;
    else
        cout << "Element not found!" << endl;

    return 0;
}

File: Java/BinarySearch.java
class BinarySearch {  

    // Binary search method
    public static void binarySearch(int arr[], int first, int last, int key) {  
        int mid = (first + last) / 2;  

        while (first <= last) {  
            if (arr[mid] < key) {  
                first = mid + 1;  // Search in the right half
            } else if (arr[mid] == key) {  
                System.out.println("Element is found at index: " + mid);  
                return;  // Element found, exit
            } else {  
                last = mid - 1;  // Search in the left half
            }  
            mid = (first + last) / 2;  
        }  

        System.out.println("Element is not found!");  // Element not present
    }  

    public static void main(String args[]) {  
        int arr[] = {10, 20, 30, 40, 50};  
        int key = 30;  
        int last = arr.length - 1;  
        binarySearch(arr, 0, last, key);     
    }  
}
